### ISSUE TITLE
compress for certain content-type (not simply true/false)

### ISSUE THREAD

**Issue by kachkaev** [containous/traefik#2124]
### Do you want to request a *feature* or report a *bug*?

feature

### What did you do?

I'm trying to configure a mix of microservices behind traefik, some of which are static websites served by nginx and some are Node.js APIs. I would like traefik to be a bit more selective about what it compresses, since now the only options are either all or nothing.

Nginx has got a nice option called [`gzip_types`](http://nginx.org/en/docs/http/ngx_http_gzip_module.html), which makes it possible to compress text responses while skipping jpegs, archives, videos and other binaries. The format is the following:

```
server {
  # ...
  gzip_types text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript image/x-icon image/bmp application/javascript application/json;
  # ...
}
```

Traefik could also use something similar in its compression middleware, i.e.:

```bash
## only certain content-types
compress = "text/plain text/css application/x-javascript text/xml application/xml application/xml+rss text/javascript image/x-icon image/bmp application/javascript application/json"
## or all
compress = true
## or nothing
compress = false
```

Being selective about what content types to compress, traefik will save CPU resources both on the server and the client while removing a need to place nginx sidecar containers next to every API endpoint that does not do compression by design (e.g. Node.js GraphQL API).

The proposed change looks backwards-compatible. WDYT?

**Comment by kachkaev** [comment]
Just came across a Node.js module called [`compressible`](https://www.npmjs.com/package/compressible), which defines content types with wildcards. So traefik option could look like:

```
compress="text/* */*+json */*+text */*+xml"
```

Or maybe use proper regexps

**Comment by kachkaev** [comment]
Just curious: what are the chances that this feature becomes a part of 1.6? 😃

**Comment by ldez** [comment]
@kachkaev not planned in 1.6 due to some design problem.

I will add this after 1.6

**Comment by kachkaev** [comment]
I'm wondering if this issue deserves to join 1.7 milestone. No worries if not, but in this case I'm curious how else to organise data compression in a micro-service environment 🤔

I see three alternative options at the moment, but all of them are pretty ugly:

1. Set `compress = false` in `traefik.toml` and make sure that each traefik-facing container compresses what it needs. This is straightforward for static nginx-based websites, but is problematic for Node.js apps. Gzipping inside a single-threaded Node.js process slows other data-processing tasks and may require extra container instances to handle the workload. Besides, someone in the team needs to be made responsible for checking that all endpoinds have compression enabled.

2. Set `compress = false` in `traefik.toml` and place sidecar containers with nginx proxies next to containers that do not support compression (for performance reasons or because their payload cannot be changed). As with option 1, someone needs to remember to always add sidecar containers to microservices that do not support gzip.

3. Set `compress = true` in `traefik.toml` and watch everything including jpegs and videos flying to clients with compression enabled. This may require significantly more resources to be allocated to traefik and also implies CPU time wasting on the client (jpegs and videos are compressed already, so gzipping them does not help). On the maintenance side, this may also cause problems like https://github.com/containous/traefik/issues/3341 & https://github.com/containous/traefik/issues/1908#issuecomment-328343800.

All three options make it quite hard to achieve a single compression policy for the entire cluster. Because of that, traefik seems to be an ideal place to configure compression for certain content-types.

**Comment by Ape** [comment]
I would also request this feature. It's not only about CPU usage or performance, since `compress = true` also drops the `Content-Length` response header. I have some big zip files available for download and would like to see the download progress percentage. I would still like to enable compression for html/css/javascript files etc.

**Comment by anjmao** [comment]
Also this is bad for images since they are already compressed. https://stackoverflow.com/questions/4957172/why-i-should-not-compress-images-in-http-headers

**Comment by kachkaev** [comment]
> if my json payload is only few kB I don't want to compress
Why would you choose not to? disabling compression by content length will add one more config option to traefik. The threshold will not be easy to reason about.

BTW content length might not be know in advance due to response streaming. This is often done to quicken the transfer of the first byte and to save microservice memory from bloating.

**Comment by ldez** [comment]
Fixed by #5721
